### DISCUSSION TITLE
how to specify the number of loop interations

### DISCUSSION THREAD

**Discussion by IngridJSJ** [Spine-project/Spine-Toolbox#1725] [Discussion]
Dear Spine team, 
I am using the loop capabilities of Spine. 
I was wondering how to stop the workflow looping according to a variable? 
Or is it possible to specify the number of times that a loop should be executed? 
Thank you!
Ingrid

**Comment by soininen** [comment]
Hi Ingrid,

Loops are controlled by Python scripts that can be entered to the editor in Loop properties dock:

![kuva](https://user-images.githubusercontent.com/19147159/174533674-990fa166-719f-40d1-be2f-15038cc3c388.png)

The script shown in the image above in fact executes the loop three times.

The *return value* of the script is used to decide if another iteration is necessary. **Zero** triggers another iteration while any other integer stops looping. You need to use Python's build-in `exit()` function to communicate the return value. `exit(0)` would loop indefinitely while `exit(1)` never loops (the items are executed exactly once).

When the script executes, it gets the value of *loop counter* as the first command line argument. You can use this value to stop the loop after a fixed number of executions. This is how it is done in the image above. Here is the same script with some comments:

```python
import sys
counter = int(sys.argv[1])  # Loop counter is supplied as first command line argument
exit(0 if counter < 3 else 1)  # Exit code 0 = continue, anything else = break
```

It might be possible to read an external variable in the script depending on where this variable is stored and use that in the stop condition. 

Loops are a bit complicated to use so let me know if you need further information!

**Comment by IngridJSJ** [comment]
Hello, 
thank you for the answer!
I have two workflow loops: 
- In the Best Investment loop I am trying to save a boolean in a csv from the previous module (tool), that is being read by the Loop and based on this boolean it continues or not . 
- In the Next year loop I am saving the simulation years to a csv and I would like the whole workflow to stop, once the final year is achieved. 
Nevertheless when the loop should continue, as the condition is not fulfilled, the next step is not executed. 
Could you give me a hint of how could this be achieved? 
![image](https://user-images.githubusercontent.com/47012366/174659451-306d252b-8624-4447-b181-139e0b33c39d.png)

**Reply by soininen** [reply]
It's difficult to say why the loop fails in your case. Is the loop script working as intended?

If 'Next year' is indeed producing an output .csv file you probably want to add that file to the Tool's output files:

![kuva](https://user-images.githubusercontent.com/19147159/174734974-19f5bc7f-a15e-4e62-b26e-947c8cdfc1bc.png)

This way the file becomes available as a command line argument to the Loop link:

![kuva](https://user-images.githubusercontent.com/19147159/174735775-0cf38ad2-e8a6-48c2-8a7e-9dcaae2b19f3.png)

The file path is then available as `sys.argv[1]` to the Loop script. See the script in the above image for an example use case.

**Reply by IngridJSJ** [reply]
Dear Antti, 
thank you for the answer. 
How can I specify the files to be saved in the current folder 
(C:\Users\...\Documents\TraderesCode\toolbox-amiris-emlab )
and not in the output results folder?
(C:\Users\...\Documents\TraderesCode\toolbox-amiris-emlab\.spinetoolbox\items\bestinvestment\output\2022-06-21T11.01.14)
Otherwise there are many files created and could be difficult to debug. 
![image](https://user-images.githubusercontent.com/47012366/174766464-b5e67b88-c930-4081-9098-8124acefcff2.png)

Currently in the "Next year" the loop when continuing (exit 0) repeats the last tool, but I would like to continue to the next module, and if the exit is 1, then to stop all simulations
In he Bestinvestment loop, if the continue (exit 0) is on, I would like the "prepare future market" tool to continue. Otherwise (exit 1), I would like the "Next year" to be executed. Is there a way to achieve this?
![tempsnip](https://user-images.githubusercontent.com/47012366/174768400-73fca0b8-340e-4182-9a9c-7c88a8b5db4f.png)


Thank you again!

**Reply by soininen** [reply]
How did you get so many `continue.txt` files on the list? That looks like a bug.

Anyway, it is not possible to specify a fixed output directory. In any case, it isn't needed as the loop script always gets path to the latest output file. Or, at least it should. Otherwise we have bug.

> Currently in the "Next year" the loop when continuing (exit 0) repeats the last tool, but I would like to continue to the next module, and if the exit is 1, then to stop all simulations

What do you mean by "repeats the last tool"? If I read your workflow correctly, continuing from Next year should rerun everything starting from decomission and including the BestInvestment-prepare future market- AMIRIS future subloop.

> In he Bestinvestment loop, if the continue (exit 0) is on, I would like the "prepare future market" tool to continue. Otherwise (exit 1), I would like the "Next year" to be executed. Is there a way to achieve this?

Isn't this already the case? Do you observe some other behavior?

**Reply by IngridJSJ** [reply]
> How did you get so many continue.txt files on the list? That looks like a bug.
The multiple years.csv and continue.txt are automatically created every time the module is executed. 

The loop repeats the last module. When I run "Next year" this step is being repeated and it seems like it takes the first "years.csv" and not the last created "years.csv"

![image](https://user-images.githubusercontent.com/47012366/174806166-5ae805e7-2b9d-4acd-bafc-dcb19edbcaff.png)
![image](https://user-images.githubusercontent.com/47012366/174807975-0056723e-e673-481c-a022-ada4ec0cd3c9.png)

![image](https://user-images.githubusercontent.com/47012366/174806278-f402dc5c-cd39-48f6-888f-602cf551fd07.png)

It could take hours to run all. I try to select "Next year" and "decomission" but it always begin executing decommission first. So do you know how to run only those two modules?

**Reply by soininen** [reply]
The fact that the loop script doesn't get the latest years.csv file is a bug. I'll investigate it.

> I try to select "Next year" and "decomission" but it always begin executing decommission first. So do you know how to run only those two modules?

decomission is always executed first because it is 'downstream' of Next year in the workflow, i.e. Next year *depends* on decomission.

I'm unsure what you mean by 'module' here. Can you clarify?

**Reply by IngridJSJ** [reply]
Well, when I select both modules (tools) -> " Next Year " and  "decomission",  "decomission" runs first. 
![image](https://user-images.githubusercontent.com/47012366/174818881-72733b1e-af5d-40a5-97a0-c745b97f6c66.png)

**Reply by IngridJSJ** [reply]
I now hard coded the output of "Next year" , and so the loop continues after "Next year"

I also tried to select more modules:
Initialize clock -> Next year -> decommission 
but when playing these tools, Initialize clock and decommission play in parallel. In other words, the tool after the loop dont wait for the loop workflow to "continue", but runs immediately and also runs after the "next year"

![image](https://user-images.githubusercontent.com/47012366/174831048-07be1c12-b4b5-4bf1-821f-13e0562dee6e.png)

**Reply by soininen** [reply]
Some good news: I managed to identify and fix the bug that was causing the Loop link to use an outdated output file from Next year. This also fixes the multiple `continue.txt` entries in Jump's Available resources. I'm unsure how you've set up Spine Toolbox, but usually you'd run `bin/upgrage_spine_reqs.bat` to apply the update.

**Reply by soininen** [reply]
> but when playing these tools, Initialize clock and decommission play in parallel. In other words, the tool after the loop dont wait for the loop workflow to "continue", but runs immediately and also runs after the "next year"

If you're trying to execute 'Next year' before decomission it just doesn't work the way your workflow is set up. This is because 'decomission' is a *predecessor* of 'Next year'. Toolbox will always execute the workflow starting from the earliest predecessor.

Would it be possible to merge decomission with Next year? I mean, if they are Python tools, would it be possible to merge the code such that Next year would both write `continue.txt` and do the decomission deeds as required? Something like this:

![kuva](https://user-images.githubusercontent.com/19147159/174971018-b3eb9205-f445-47b5-82c3-cccc4324d686.png)

**Reply by IngridJSJ** [reply]
Hello Antti, 
Thank you very much for solving that bug. I will try it now. 
> Would it be possible to merge decomission with Next year? 

Yes I can but I would still have the same problem. 
I tried to just erase the decommission step for now, but the "Initialize Clock" and "Prepare Data Amiris" are starting at the same time. 
And when I select both Next year and  "Prepare Data Amiris", the  "Prepare Data Amiris" starts first
![image](https://user-images.githubusercontent.com/47012366/174973274-217f4759-b70a-43ec-81c3-4dfec010b8cd.png)
How can I specify which tool is predecessor of which tool?

**Reply by soininen** [reply]
> I tried to just erase the decommission step for now, but the "Initialize Clock" and "Prepare Data Amiris" are starting at the same time.
And when I select both Next year and "Prepare Data Amiris", the "Prepare Data Amiris" starts first

It is the same issue that you had with 'decomission': 'Prepare Data Amiris' and 'Initialize Clock' are both predecessors to 'Next year'. Toolbox will execute them before 'Next year' no matter what. Since 'Prepare Data Amiris' and 'Initialize Clock' are in two different branches in the workflow leading eventually to 'Next year', Toolbox executes them parallel.

If you really want to start execution from 'Next year', you need make it a predecessor to other items in the workflow. Perhaps duplicating (Ctrl+d) the item like this would do the trick:

![kuva](https://user-images.githubusercontent.com/19147159/174982452-0e061c33-b680-4464-a050-b449d1eee9d9.png)


Now, if you select 'Initial next year', 'Prepare Data Amiris' and 'Next year' from the workflow above, Toolbox will execute 'Initial next year' before 'Prepare Data Amiris' and then loop between 'Next year' and 'Prepare Data Amiris' until the Loop condition is satisfied.

**Reply by IngridJSJ** [reply]
hi Antii, 

so I now made the Initialize clock before decommission and that worked well. 
The complete workflow is very long so I made a quick test and it seems like it will work, once the whole workflow is complete.
![image](https://user-images.githubusercontent.com/47012366/174995192-aacd3269-6032-4cd7-9c68-9bee8690fe5e.png)

Now it seems that the numerations in the other part of the workflow are messed up.
![image](https://user-images.githubusercontent.com/47012366/174997822-0d51d00d-0951-44c4-9a33-8e7781969dd6.png)

But I think actually the problem is that the number displayed in the top left icons are not showing double digits. 
I hope that the tools will play in the right order, but will only be able to test later, once all tools are working, 
Anyways, thank you for the support! 


BTW that sudden black kitty was so surprising, a bit confusing but very cool!!!

**Reply by soininen** [reply]
> so I now made the Initialize clock before decommission and that worked well. The complete workflow is very long so I made a quick test and it seems like it will work, once the whole workflow is complete.

Good to know! I haven't seen such complex workflows before - you're really stretching Toolbox here.

I also believe the little box on the upper left corner is just too small to show double digits. It's a visual bug that needs to be fixed at some point.

> BTW that sudden black kitty was so surprising, a bit confusing but very cool!!!

Haha, the cat sometimes surprises me, too!
